Mathlib/FieldTheory/CardinalEmb.lean

```diff
@@ -0,0 +1,344 @@
+to the initial ordinal `ι` of the cardinal `Module.rank F E`, because the dimension of the
+subalgebra generated by an infinite set cannot be greater than the cardinality of the set, and
+in an algebraic extension, any subalgebra is a field. This is proven as `Algebra.rank_adjoin_le`
+and used to show `leastExt` is a total function from `ι` to itself. It is probably
+mathematically the most nontrivial part of the whole argument, but turned out easy to
+formalize and was done the earliest.
+
+Once we have the filtration `E⟮<i⟯` for `i : ι`, we may build an embedding `E →ₐ[F] Ē` step by
+step. To extend an embedding `E⟮<i⟯ →ₐ[F] Ē` to the successor `E⟮<i⁺⟯`, the number of choices
+is equal to the (finite) separable degree of `E⟮<i⁺⟯ / E⟮<i⟯`, which is equal to its rank if
+`E/F` is separable. Since each extension is nontrivial, the degree is at least two (`two_le_deg`)
+but always finite. Intuitively, these choices multiply together to give the cardinality of
+`Field.Emb F E := (E →ₐ[F] Ē)`, and since the total times of choices to be made is the length of
+the filtration `#ι`, we conclude that `2 ^ #ι ≤ #(Field.Emb F E) ≤ ℵ₀ ^ #ι`, but for infinite `#ι`
+both sides are equal, so we get an equality `#(Field.Emb F E) = 2 ^ #ι = 2 ^ Module.rank F E`.
+
+To rigorize the argument we formalize the choice at step `i` as a bijection `F i⁺ ≃ F i × X i`,
+where `X i := Field.Emb E⟮<i⟯ E⟮<i⁺⟯`, and we formalize the combination of all choices
+as the Pi type `∀ i : ι, X i`. We use transfinite recursion (`SuccOrder.prelimitRecOn`)
+to build a bijection `Field.Emb F E ≃ ∀ i, X i` with the Pi type by successively extending
+bijections `F i ≃ ∀ j : Iio i, X j` using the bijections `F i⁺ ≃ F i × X i` with product types
+(`InverseSystem.piEquivSucc`). More details are found in the file about `InverseLimit`.
+Since `ι` is a limit ordinal, `Field.Emb F E ≃ (⊤ →ₐ[F] Ē)` is not actually one of the `F i`
+because `⊤` is not one of the `E⟮<i⟯`, so we have to adjoin a top element to `ι` (`WithTop ι`)
+to obtain the bijection
+`Field.Emb F E ≃ F ⊤ ≃ ∀ j : Iio ⊤, X j ≃ ∀ i : ι, X i = ∀ i : ι, Field.Emb E⟮<i⟯ E⟮<i⁺⟯`.
+To make this straightforward, it is crucial that `(↑i : WithTop ι)⁺ = ↑(i⁺)` holds definitionally.
+
+The predicate `IsSuccPrelimit` allows us to treat limits and the bottom element uniformly, and
+the only place the bottom element requires special treatment is in `equivLim` (the bijection
+between `E⟮<i⟯ →ₐ[F] Ē` and the inverse limit of `E⟮<j⟯ →ₐ[F] Ē` over `j < i`).
+
+-/
+
+open Cardinal Module.Free Set Order IntermediateField InverseSystem
+
+universe u v
+
+variable (F : Type u) (E : Type v) [Field F] [Field E] [Algebra F E]
+
+namespace Field.Emb
+
+namespace Cardinal
+
+noncomputable section
+
+set_option quotPrecheck false
+
+/-- Index a basis of E/F using the initial ordinal of the cardinal `Module.rank F E`. -/
+local notation "ι" => (Module.rank F E).ord.toType
+
+private local instance : SuccOrder ι := SuccOrder.ofLinearWellFoundedLT ι
```

Comment:
Does this private local instance used in the *declarations* of the public theorems in this file? If yes, then these results can't be used by other files due to missing instance.

There are also some other private instances below.

Reply:
I think as long as the types of declarations don't require the instance, they can still be used outside of this file, though I don't expect them to be. My main motivation for not making them private is so that they show up in the documentation, so the reader gets a flavor of the proof without reading the code; otherwise the documentation will only contain the last 5 theorems (the main results). I've enclosed everything other than the main results in the namespace `Field.Emb.Cardinal` that probably will never be opened, so they are effectively private.

Reply:
There are only a few declarations whose types mention `succ`: `filtration_succ, succEquiv, succEquiv_coherence, equivSucc, equivSucc_coherence`. Even these won't stop to work in a new file, as they'll carry the private local instance with them. There are only one more private local instance (`Decidable`) in this file, and aother non-private local instance (`NoMaxOrder`). The main reason for making the `SuccOrder` and `Decidable` instances private is that they're type-valued and I don't want to add docstrings to them. I can certainly make them non-private and add docstrings if you want.

Reply:
> I can certainly make them non-private and add docstrings if you want.

No, I have no preferences on them. Feel free to decide by yourself.